src/failure_detector/failure_detector.cpp

```diff
@@ -415,33 +431,47 @@ bool failure_detector::end_ping_internal(::dsn::error_code err, const beacon_ack
     /*
      * the caller of the end_ping_internal should lock necessarily!!!
      */
+    host_port hp_this_node, hp_primary_node;
+    if (ack.__isset.hp_this_node) {
+        hp_this_node = ack.hp_this_node;
+    } else {
+        hp_this_node = host_port(ack.this_node);
+    }
+    if (ack.__isset.hp_primary_node) {
+        hp_primary_node = ack.hp_primary_node;
+    } else {
+        hp_primary_node = host_port(ack.primary_node);
+    }
+
     uint64_t beacon_send_time = ack.time;
-    auto node = ack.this_node;
 
     if (err != ERR_OK) {
         LOG_WARNING("ping master({}) failed, timeout_ms = {}, err = {}",
-                    node,
+                    hp_this_node,
                     _beacon_timeout_milliseconds,
                     err);
         _recent_beacon_fail_count->increment();
     }
 
-    master_map::iterator itr = _masters.find(node);
+    master_map::iterator itr = _masters.find(hp_this_node);
 
     if (itr == _masters.end()) {
         LOG_WARNING("received beacon ack without corresponding master, ignore it, "
-                    "remote_master[{}], local_worker[{}]",
-                    node,
+                    "remote_master[{}({})], local_worker[{}]",
+                    hp_this_node,
+                    ack.this_node,
                     dsn_primary_address());
```

Comment:
Also logging host_port by dsn_primary_host_port().

Other places are the same.